contributors/design-proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,166 @@
+## Refactor Cloud Provider out of Kubernetes Core
+
+As kubernetes has evolved tremendously, it has become difficult for different cloudproviders (currently 7) to make changes and iterate quickly. Moreover, the cloudproviders are constrained by the kubernetes build/release life-cycle. It has also become harder to add new cloud providers. This proposal aims to move towards a kubernetes code base where cloud providers specific code will move out of the core repository and into "official" repositories, where it will be maintained by the cloud providers themselves.
+
+### 1. Current use of Cloud Provider
+
+The following components have cloudprovider dependencies
+
+    1. kube-controller-manager
+    2. kubelet
+    3. kube-apiserver
+
+#### Cloud Provider in Kube-Controller-Manager
+
+The kube-controller-manager has many controller loops
+
+ - nodeController
+ - volumeController
+ - routeController
+ - serviceController
+ - replicationController
+ - endpointController
+ - resourceQuotaController
+ - namespaceController
+ - deploymentController
+ - etc..
+
+Among these controller loops, the following are cloud provider dependent.
+
+ - nodeController
+ - volumeController
+ - routeController
+ - serviceController
+
+The nodeController uses the cloudprovider to check if a node has been deleted from the cloud. If cloud provider reports a node as deleted, then this controller immediately deletes the node from kubernetes. This check removes the need to wait for a specific amount of time to conclude that an inactive node is actually dead.
+
+The volumeController uses the cloudprovider to create, delete, attach and detach volumes to nodes. For instance, the logic for provisioning, attaching, and detaching a EBS volume resides in the AWS cloudprovider. The volumeController uses this code to perform its operations.
+
+The routeController configures routes for hosts in Google Cloud.
+
+The serviceController maintains a list of currently active nodes, and is responsible for creating and deleting LoadBalancers in the underlying cloud.
+
+#### Cloud Provider in Kubelet
+
+Moving on to the kubelet, the following cloud provider dependencies exist in kubelet.
+
+ - Find the cloud nodename of the host that kubelet is running on for the following reasons :
+      1. To obtain the config map for the kubelet, if one already exists
+      2. To uniquely identify current node using nodeInformer
+      3. To instantiate a reference to the current node object
+ - Find the InstanceID, ProviderID, ExternalID, Zone Info of the node object while initializing it
+ - Periodically poll the cloud provider to figure out if the node has any new IP addresses associated with it
+ - In case of Google cloud, it allows the provider to configure routes for the node
```

Comment:
This is unclear.  Kubelet doesn't have anything to do with the routes itself, it leaves all that to the route provider.  What kubelet does (wrongly, see https://github.com/kubernetes/kubernetes/pull/34398) is set the "NetworkUnavailable" condition whenever a cloud provider is specified.  That's incorrect because the only thing that clears this condition is the route controller, and the route controller is only required if the networking setup uses direct routing between nodes for the pod network, which many plugins do not (including those with overlays).

In any case, I'd clarify to say something like "Assumes cloud providers require routes for network plugins (not always true) and blocks readiness until cloud routes are configured"